File: src/main/java/io/femo/http/transport/http2/frames/RstStreamFrame.java
package io.femo.http.transport.http2.frames;

import io.femo.http.Constants;
import io.femo.http.transport.http2.HttpFrame;
import io.femo.http.transport.http2.HttpFrameException;
import io.femo.http.transport.http2.HttpSettings;
import io.femo.http.transport.http2.HttpUtil;

/**
 * Created by felix on 9/18/16.
 */
public class RstStreamFrame extends HttpFrame {

    private int errorCode;

    /**
     * Constructs a new HttpFrame with no payload and information
     *
     * @param settings the settings for the current HTTP/2,0 connection
     */
    public RstStreamFrame(HttpSettings settings) {
        this(settings, Constants.Http20.ErrorCodes.NO_ERROR);
    }

    public RstStreamFrame(HttpSettings settings, int errorCode) {
        super(settings);
        setType(Constants.Http20.FrameType.RST_STREAM);
    }

    public int getErrorCode() {
        return errorCode;
    }

    public void setErrorCode(int errorCode) {
        this.errorCode = errorCode;
    }

    @Override
    public void build() {
        setPayload(HttpUtil.toByte(errorCode));
    }

    public static RstStreamFrame from(HttpFrame frame) {
        if(frame.getLength() != 4) {
            throw new HttpFrameException("Invalid length for RST_STREAM frame");
        }
        if(frame.getType() != Constants.Http20.FrameType.RST_STREAM) {
            throw new HttpFrameException("Invalid type for RST_STREAM frame");
        }
        RstStreamFrame rstStreamFrame = new RstStreamFrame(frame.getSettings());
        rstStreamFrame.setStreamIdentifier(frame.getStreamIdentifier());
        rstStreamFrame.setFlags(frame.getFlags());
        rstStreamFrame.setErrorCode(HttpUtil.toInt(frame.getPayload()));
        return rstStreamFrame;
    }
}

File: src/main/java/org/web25/http/transport/http2/frames/RstStreamFrame.kt
package org.web25.http.transport.http2.frames

import org.web25.http.Constants
import org.web25.http.transport.http2.HttpFrame
import org.web25.http.transport.http2.HttpFrameException
import org.web25.http.transport.http2.HttpSettings
import org.web25.http.transport.http2.HttpUtil

/**
 * Created by felix on 9/18/16.
 */
class RstStreamFrame @JvmOverloads constructor(settings: HttpSettings, var errorCode: Int = Constants.Http20.ErrorCodes.NO_ERROR) : HttpFrame(settings) {

    init {
        type = Constants.Http20.FrameType.RST_STREAM
    }

    override fun build() {
        payload = HttpUtil.toByte(errorCode)
    }

    companion object {

        fun from(frame: HttpFrame): RstStreamFrame {
            if (frame.length != 4) {
                throw HttpFrameException("Invalid length for RST_STREAM frame")
            }
            if (frame.type != Constants.Http20.FrameType.RST_STREAM) {
                throw HttpFrameException("Invalid type for RST_STREAM frame")
            }
            val rstStreamFrame = RstStreamFrame(frame.settings)
            rstStreamFrame.streamIdentifier = frame.streamIdentifier
            rstStreamFrame.flags = frame.flags
            rstStreamFrame.errorCode = HttpUtil.toInt(frame.payload)
            return rstStreamFrame
        }
    }
}
/**
 * Constructs a new HttpFrame with no payload and information

 * @param settings the settings for the current HTTP/2,0 connection
 */
